gateway-sp-comms/src/shared_socket.rs

```diff
@@ -78,67 +81,89 @@ pub struct BindError {
 /// the packet). If it matches an interface that has a `SingleSp` handler, the
 /// message is forwarded to that `SingleSp` instance via a tokio channel;
 /// otherwise, the packet is discarded.
-#[derive(Debug)]
-pub struct SharedSocket {
+pub struct SharedSocket<T: Send> {
     socket: SendOnlyUdpSocket,
     scope_id_cache: Arc<ScopeIdCache>,
-    single_sp_handlers:
-        Arc<Mutex<FxHashMap<Name, mpsc::Sender<SingleSpMessage>>>>,
+    single_sp_handlers: SpHandlerMap<T>,
     recv_handler_task: JoinHandle<()>,
     log: Logger,
 }
 
-impl Drop for SharedSocket {
+// Hand-rolled `Debug` impl as the message type (`T`) needn't be `Debug` for the
+// `SharedSocket` to be debug.
```

Comment:
This seems fine, but I'll plug https://docs.rs/derive-where/latest/derive_where/ since we use it in omicron, if you want to pull it in here too.